text/0000-grid-identity.md

```diff
@@ -0,0 +1,718 @@
+function, which is provided to the smart contract by the Grid SDK. The
+implementation of `has_permission()` makes various calls to obtain portions of
+state and thus requires a handle to state be passed into the function.
+
+## Understanding permission checking in Pike
+
+The current Pike implementation relies on organizations, agents, and roles:
+
+![Pike Diagram](https://github.com/Cargill/grid-rfcs/blob/identity/images/grid-identity/pike.png)
+
+In Pike, when `has_permission()` is evaluated, an agent is looked up via the
+public key. The public key is the agent’s primary key and the agent is stored in
+state using an address derived from the public key. The agent has a list of
+roles, which are strings. The `has_permission()` function returns true if the
+`permission_name` is in the list of roles and false if it is not.
+
+Each Pike agent must be part of an organization. For many smart contracts in
+Grid, the records created by smart contracts also include the ID of the
+organization that they are associated with. For instance, a product record
+includes the Pike organization ID of the organization that owns the product
+record. In order to ensure that agents can't create unauthorized records on
+behalf of an organization they are not a part of, smart contracts have to check
+the organization ID in the record matches up with the agent's organization ID.
+If not, the transaction is considered invalid.
+
+This pattern is effective in preventing agents from acting on behalf of other
+organizations. However, there may be cases where an organization needs to grant
+certain permissions to another organization. With Pike this would not be
+possible without specific workaround in each smart contract. This could lead to
+inconsistency and sub-standard implementations across Grid. It is imperative
+that Grid Identity solves this problem.
+
+## Redefining Roles
+
+In Grid Identity, roles are not directly used as permissions. Instead, a role is
+defined as a list of permissions. The smart contracts check for permissions.
+This level of indirection allows the definition of higher-order roles be defined
+by organizations. The resulting state looks like:
+
+![Grid Identity Diagram](https://github.com/Cargill/grid-rfcs/blob/identity/images/grid-identity/identity.png)
+
+The Agent has one or more organization-specific roles. The `has_permission()`
+function determines the roles assigned to the agent and returns `true` if one of
+the roles contains `permission_name` or `false` otherwise.
+
+## Delegation of Roles Between Organizations
+
+In order to support delegation between organizations, a role has fields called
+`allowed_organizations` and `inherit_from`.
+
+The `allowed_organizations` field is a list of organizations other than the
+defining organization who can use the permissions granted in the role. This
+allows organizations to delegate roles to each other.
+
+The `inherit_from` field is a list containing references to other roles. If a
```

Comment:
Adding a comment here to discuss merits of the feature. It was brought up in the Grid community meeting as well. Would it be of help to stop inheriting certain permissions?

Reply:
I could see this potentially being useful but I would like to see a specific example where we need it. Since an org has full control over what roles they define and what permissions they delegate, im not sure if that provides any unique functionality. I could be missing something though.